src/linter/root.go

```diff
@@ -829,30 +829,83 @@ func (d *rootWalker) enterPropertyList(pl *ir.PropertyListStmt) bool {
 	}
 
 	for _, pNode := range pl.Properties {
-		p := pNode.(*ir.PropertyStmt)
+		prop := pNode.(*ir.PropertyStmt)
 
-		nm := p.Variable.Name
+		nm := prop.Variable.Name
 
-		if p.Expr != nil {
-			phpDocType = phpDocType.Append(solver.ExprTypeLocal(d.scope(), d.ctx.st, p.Expr))
+		// We need to clone the types, because otherwise, if several
+		// properties are written in one definition, and null was
+		// assigned to the first, then all properties become nullable.
+		propTypes := phpDocType.Clone().Append(typeHintType)
+
+		d.checkAssignNullToNotNullableProperty(prop, propTypes)
+
+		if prop.Expr != nil {
+			propTypes = propTypes.Append(solver.ExprTypeLocal(d.scope(), d.ctx.st, prop.Expr))
 		}
-		phpDocType = phpDocType.Append(typeHintType)
 
 		if isStatic {
 			nm = "$" + nm
 		}
 
 		// TODO: handle duplicate property
 		cl.Properties[nm] = meta.PropertyInfo{
-			Pos:         d.getElementPos(p),
-			Typ:         phpDocType.Immutable(),
+			Pos:         d.getElementPos(prop),
+			Typ:         propTypes.Immutable(),
 			AccessLevel: accessLevel,
 		}
 	}
 
 	return true
 }
 
+func (d *rootWalker) checkAssignNullToNotNullableProperty(prop *ir.PropertyStmt, propTypes types.Map) {
+	assignNull := false
+
+	if expr, ok := prop.Expr.(*ir.ConstFetchExpr); ok {
+		assignNull = expr.Constant.Value == "null"
+	}
+
+	if assignNull && !propTypes.Empty() {
+		onlyClasses := true
```

Comment:
I would add a test with something like `int|Foo` where we mix classes and non-classes.

Reply:
Added.